### ISSUE TITLE
`Parameter "**kwargs" has no corresponding parameter` with generic `ParamSpec`

### ISSUE THREAD

**Issue by multimeric** [microsoft/pyright#3654]
**Describe the bug**
When using a decorator function which is generic over the `ParamSpec` of the inner function, you will get the error: `Parameter "**kwargs" has no corresponding parameter (reportGeneralTypeIssues)`

**To Reproduce**
Here is a minimal reproduction. You can tweak a lot of the code and it will still raise this error, as long as `decorator` remains generic, using a generic `ParamSpec`.
```python
from typing import Callable, Any, ParamSpec

Params = ParamSpec("Params")
Call = Callable[Params, Any]

def decorator(func: Call) -> Call:
    return func

@decorator
def foo():
    pass
```

**Expected behavior**
I don't think this should not throw an error. The only thing I'm unsure about is whether it's valid to use `Call` as though it's a generic `TypeVar` when it isn't exactly that. If my usage is invalid, then it would be better for this warning to explain that, rather than throw this misleading error about `**kwargs`.

**VS Code extension or command-line**
* Running via CLI (e.g. `pyright test.py`)
* pyright `1.1.257`
* Python `3.10.4`

**Comment by erictraut** [comment]
You have defined a generic type alias `Call` that accepts a single type argument, but when you are using this type alias, you are not supplying any type argument. I suspect what you meant here was:

```python
def decorator(func: Call[Params]) -> Call[Params]:
    return func
```

**Comment by erictraut** [comment]
If you don't provide a type argument, `Call` should default to `Call[...]`. In other words, `Call` should behaviorally act just like `Call[...]`, but the behavior currently differs. That is a bug, and I've introduced a fix that will be included in the next release.

That said, I'm guessing that it is not your intent to use `Call[...]` here. Normally decorators preserve the call signature of the function they are decorating, so you probably want to use `Call[ParamSpec]` in your decorator signature.

**Comment by multimeric** [comment]
Thanks for the fix. Unfortunately I still don't understand what I'm doing wrong here. Why would I need to use `Call[Params]` each time, when I've already provided the type variable `Params`? Nevertheless, it's not your job to explain that to me. If you believe this is the correct behaviour, and the error message is now fixed, then I'm happy to close.

**Comment by erictraut** [comment]
I'm happy to explain. 

In Python, there is no dedicated syntax for declaring a type alias. In most (all?) other modern language there is. (FWIW, I just submitted a [new PEP](https://peps.python.org/pep-0695/) that proposes to fix this Python 3.12, but it is in the early stages of the standardization process.)

In Python today, a type alias declaration uses the same syntax as a simple variable assignment. The expression on the right-hand side (RHS) indicates the type. If the expression on the RHS uses one or more type variables or param specs, the type alias is generic.

For example:

```python
# This generic type alias has one type parameter T
ListOrSet = list[T] | set[T]

# This generic type alias has two type parameters K and V
SpecialDict = dict[K, V]
```

Generic type aliases can be used like generic classes within type annotations. When they are used, type arguments must be supplied for the associated parameters. For example:

```python
a: ListOrSet[int] = {1, 2, 3}
b: SpecialDict[str, float] = {"x": 1.0}
```

As with generic classes, if you omit the type arguments, a type checker will fill in the missing type arguments with `Any`. Or in the case of a ParamSpec, it will fill in `...`, which is the equivalent of `Any` for a ParamSpec.

In your code snippet above, you define a generic type alias called `Call`. It is parameterized by a single ParamSpec.

```python
Call = Callable[Params, Any]
```

Any place you use this type alias, you should provide a type argument for the ParamSpec.

```python
def func1(x: int): ...
x: Call[int] = func1

def func2(x: str, y: int): ...
y: Call[str, int] = func2
```

You can also use `...` as the argument for a ParamSpec. 

```python
x: Call[...] = func1
```

If you omit the type argument altogether, a type checker will assume a type argument of `...`. So your decorator was effectively annotated as follows:

```python
def decorator(func: Call[...]) -> Call[...]:
    return func
```

That is probably not what you meant because it hides the signature of the decorated function. What you probably wanted was to preserve the signature of the decorated function. That requires you to use a ParamSpec as the type argument for the `Call` type alias.

```python
def decorator(func: Call[Params]) -> Call[Params]:
    return func
```

The current syntax for type parameters is very confusing in Python. That's why there's an effort underway to improve the syntax. If PEP 695 is approved in its current form, you will be able to write your original code snippet as follows:

```python
from typing import Callable, Any

type Call[**P] = Callable[P, Any]  # New syntax for declaring a generic type alias

def decorator[**P](func: Call[P]) -> Call[P]: # New syntax for declaring a generic function
    return func
```

Or if you want to preserve the return type of the decorated function...

```python
from typing import Callable

def decorator[**P, R](func: Callable[P, R]) -> Callable[P, R]:
    return func
```

**Comment by erictraut** [comment]
Another common approach for annotating decorators is to use a normal TypeVar that is bound to a callable.

```python
T = TypeVar("T", bound=Callable[..., Any])
def decorator(func: T) -> T:
    return func
```

**Comment by multimeric** [comment]
Right, I think I see now. I guess this behaviour allows the `TypeVar`s within a `TypeAlias` to be bound to a concrete type in certain cases which can be helpful. In my case though I want them to remain generic, which is why it was counterintuitive to me that I should have to re-specify the `TypeVar`s. It is helpful to know that this concept is called a `TypeAlias`, which has led me to find the PEP that documents this behaviour (although in minimal detail, I would say): https://peps.python.org/pep-0484/#type-aliases.

I love the syntax for introducing new generic types in your PEP, but I wasn't aware there was also a new syntax for type aliases in there.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.258, which I just published. It will also be included in the next release of pylance.
